### PR TITLE
[MM-25186] Bot username validation i8n support

### PR DESCRIPTION
**Pull request by prithvijit-dasgupta**
#### Summary

This PR addresses the following for the `add_bot` and `utils` module
1. Uses i8n code for Bot username validation errors
2. Minor refactor of `utils.isValidUserName`
3. Add test cases for `isValidUsername` and `isValidBotUsername` 

#### Ticket Link

Jira ticket: https://mattermost.atlassian.net/browse/MM-25186
Fixes https://github.com/mattermost/mattermost-server/issues/15555

### CODE REVIEW

**Comment by prithvijit-dasgupta** [comment]
Also minor thing but `webapp` repo doesn't have the `hacktoberfest` topic added
Screenshot 2020-10-05 at 2 35 44 PM

**Comment by jasonblais** [comment]
> Also minor thing but webapp repo doesn't have the hacktoberfest topic added

Thanks! We're making changes soon to ensure our repositories have the `hacktoberfest` topic and PRs have the `hacktoberfest-accepted` label.

Also thanks for contributing :)

**Review by aaronrothschild — APPROVED** [review]
Thanks for your contribution @prithvijit-dasgupta ! Looks good to me!

File: components/integrations/bots/add_bot/add_bot.jsx [hunk 1]
```diff
@@ -285,9 +285,26 @@ export default class AddBot extends React.PureComponent {
...
+                }
                 this.setState({
                     adding: false,
-                    error: usernameError,
+                    error: (
+                        <FormattedMessage
+                            id='bots.manage.add.invalid_username'
+                            defaultMessage='Usernames have to begin with a lowercase letter and be 3-22 characters long. You can use lowercase letters, numbers, periods, dashes, and underscores.'
+                        />
+                    ),
```
Comment (iomodo) [inline components/integrations/bots/add_bot/add_bot.jsx:307]: Depending on the error we could create a appropriate `FormattedMessage` first and then call the `this.setState` method.

**Review by iomodo — CHANGES_REQUESTED** [review]
Thanks for the contribution @prithvijit-dasgupta. Looks good! Minor changes below. I was just wondering should we be checking for different errors like `INVALID_CHARACTERS`, `INVALID_LENGTH`, etc and outputting the appropriate message?
@aaronrothschild  could you provide some clarity here?

File: components/integrations/bots/add_bot/add_bot.jsx [hunk 2]
```diff
@@ -285,9 +285,26 @@ export default class AddBot extends React.PureComponent {
         } else {
             const usernameError = Utils.isValidBotUsername(bot.username);
             if (usernameError) {
+                if (usernameError.id === 'INVALID_LAST_CHARACTER') {
```
[the remark this answers is not in the document]
Comment (prithvijit-dasgupta) [reply components/integrations/bots/add_bot/add_bot.jsx:288]: @iomodo Should I only do it specifically for `INVALID_LAST_CHARACTER` in the `add_bot` module?

**Comment by prithvijit-dasgupta** [comment]
Waiting for some inputs from @aaronrothschild on error messages.

File: components/integrations/bots/add_bot/add_bot.jsx [hunk 3]
```diff
@@ -285,9 +285,26 @@ export default class AddBot extends React.PureComponent {
...
+                    this.setState({
+                        adding: false,
+                        error: (
+                            <FormattedMessage
+                                id='bots.manage.add.invalid_last_char'
+                                defaultMessage='Bot usernames cannot have period as last character'
```
Comment (aaronrothschild) [inline components/integrations/bots/add_bot/add_bot.jsx:294]:
```suggestion
                                defaultMessage='Bot usernames cannot have a period as the last character'
```

File: components/integrations/bots/add_bot/add_bot.jsx [hunk 4]
```diff
@@ -285,9 +285,26 @@ export default class AddBot extends React.PureComponent {
...
+                    });
+                    return;
+                }
                 this.setState({
                     adding: false,
-                    error: usernameError,
+                    error: (
+                        <FormattedMessage
+                            id='bots.manage.add.invalid_username'
+                            defaultMessage='Usernames have to begin with a lowercase letter and be 3-22 characters long. You can use lowercase letters, numbers, periods, dashes, and underscores.'
```
Comment (aaronrothschild) [inline components/integrations/bots/add_bot/add_bot.jsx:305]:
```suggestion
                            defaultMessage='Usernames must begin with a lowercase letter and be 3-22 characters long. You can use lowercase letters, numbers, periods, dashes, and underscores.'
```

File: i18n/en.json [hunk 5]
```diff
@@ -2122,6 +2122,7 @@
   "bots.manage.add.cancel": "Cancel",
   "bots.manage.add.create": "Create Bot Account",
   "bots.manage.add.creating": "Creating...",
+  "bots.manage.add.invalid_last_char": "Bot usernames cannot have period as last character",
```
Comment (aaronrothschild) [inline i18n/en.json:2125]:
```suggestion
  "bots.manage.add.invalid_last_char": "Bot usernames cannot have a period as the last character",
```

**Review by aaronrothschild — APPROVED** [review]
Small changes

File: components/integrations/bots/add_bot/add_bot.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (iomodo) [reply components/integrations/bots/add_bot/add_bot.jsx:288]: @prithvijit-dasgupta all strings should probably be consts, same goes to `INVALID_LENGTH`, `INVALID_CHARACTERS`, ...

[the remark this answers is not in the document]
Comment (prithvijit-dasgupta) [reply components/integrations/bots/add_bot/add_bot.jsx]: done. Created new constant `ValidationErrors`

File: components/integrations/bots/add_bot/add_bot.jsx (hunk 1 above)
Comment (prithvijit-dasgupta) [reply components/integrations/bots/add_bot/add_bot.jsx]: done

File: components/integrations/bots/add_bot/add_bot.jsx (hunk 3 above)
Comment (prithvijit-dasgupta) [reply components/integrations/bots/add_bot/add_bot.jsx]: done

File: components/integrations/bots/add_bot/add_bot.jsx (hunk 4 above)
Comment (prithvijit-dasgupta) [reply components/integrations/bots/add_bot/add_bot.jsx]: done

File: i18n/en.json (hunk 5 above)
Comment (prithvijit-dasgupta) [reply i18n/en.json]: done

**Review by aaronrothschild — APPROVED** [review]
Great stuff, thanks for you contribution @prithvijit-dasgupta !

**Review by larkox — APPROVED** [review]
LGTM! Thanks! :tada:

**Review by iomodo — APPROVED** [review]
LGTM! Thanks for the contribution @prithvijit-dasgupta !

**Comment by jfrerich** [comment]
@prapti, reminder for review.  

Thanks for the contribution, @prithvijit-dasgupta!  QA is likely very busy with hacktoberfest, but someone will get to your review shortly!

**Review by prapti — APPROVED** [review]
Tested on the PR server. Changes look good. The error for invalid bot username gets translated as expected.
Approving.
Thank you so much for your contribution @prithvijit-dasgupta! 🎉

**Comment by hanzei** [comment]
Thanks for the contribution @prithvijit-dasgupta :+1:

**Pull request merged**